tezos/new_context/src/working_tree/storage.rs

```diff
@@ -0,0 +1,574 @@
+    Ref { start: usize, end: usize },
+}
+
+impl BlobStorageId {
+    pub fn new_inline(value: &[u8]) -> Result<Self, StorageIdError> {
+        let len = value.len();
+
+        // Inline values are 7 bytes maximum
+        if len > 7 {
+            return Err(StorageIdError::BlobSliceTooBig);
+        }
+
+        // We copy the slice into an array so we can use u64::from_ne_bytes
+        let mut new_value: [u8; 8] = [0; 8];
+        new_value[..len].copy_from_slice(value);
+        let value = u64::from_ne_bytes(new_value);
+
+        let blob_id = Self {
+            bits: (1 << 60) | (len as u64) << 56 | value,
+        };
+
+        debug_assert_eq!(
+            blob_id.get(),
+            BlobRef::Inline {
+                length: len.try_into().unwrap(),
+                value: new_value[..7].try_into().unwrap()
+            }
+        );
+
+        Ok(blob_id)
+    }
+
+    fn new(start: usize, end: usize) -> Result<Self, StorageIdError> {
+        let length = end - start;
+
+        if start & !0xFFFFFFFF != 0 {
+            // Start must fit in 32 bits
+            return Err(StorageIdError::BlobStartTooBig);
+        }
+
+        if length & !0xFFFFFFF != 0 {
+            // Length must fit in 28 bits
+            return Err(StorageIdError::BlobLengthTooBig);
+        }
+
+        let blob_id = Self {
+            bits: (start as u64) << 28 | length as u64,
+        };
+
+        debug_assert_eq!(blob_id.get(), BlobRef::Ref { start, end });
+
+        Ok(blob_id)
+    }
+
+    fn get(self) -> BlobRef {
+        if self.is_inline() {
+            let length = ((self.bits >> 56) & 0xF) as u8;
+
+            // Extract the inline value and make it a slice
+            let value: u64 = self.bits & 0xFFFFFFFFFFFFFF;
+            let value: [u8; 8] = value.to_ne_bytes();
+            let value: [u8; 7] = value[..7].try_into().unwrap(); // Never fails, `value` is [u8; 8]
+
+            BlobRef::Inline { length, value }
+        } else {
+            let start = (self.bits >> 28) as usize;
+            let length = (self.bits & 0xFFFFFFF) as usize;
+
+            BlobRef::Ref {
+                start,
+                end: start + length,
+            }
+        }
+    }
+
+    pub fn is_inline(self) -> bool {
+        self.bits >> 60 != 0
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct NodeId(u32);
+
+#[derive(Debug)]
+pub struct NodeIdError;
+
+impl TryInto<usize> for NodeId {
+    type Error = NodeIdError;
+
+    fn try_into(self) -> Result<usize, Self::Error> {
+        Ok(self.0 as usize)
+    }
+}
+
+impl TryFrom<usize> for NodeId {
+    type Error = NodeIdError;
+
+    fn try_from(value: usize) -> Result<Self, Self::Error> {
+        value.try_into().map(NodeId).map_err(|_| NodeIdError)
+    }
+}
+
+#[derive(Debug)]
+pub struct StorageMemoryUsage {
+    node_cap: usize,
+    trees_cap: usize,
+    temp_tree_cap: usize,
+    blobs_cap: usize,
+    strings: StringsMemoryUsage,
+}
+
+pub struct Storage {
```

Comment:
This one needs a comment explaining its purpose and where it is used. I now have the impression that it is used for keeping the working tree data (because it is cleared after the commit), but I am still not sure I fully understand it.